src/library/basm.c

```diff
@@ -1728,38 +1731,187 @@ void basm_save_to_file_as_nasm(Basm *basm, const char *output_file_path)
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rax, [rsi]\n");
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rbx, [rsi]\n");
+            fprintf(output, "    xor rax, rbx\n");
+            fprintf(output, "    mov [rsi], rax\n");
+            fprintf(output, "    add rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov [stack_top], rsi\n");
+        }
+        break;
+        case INST_SHR: {
+            fprintf(output, "    ;; shr\n");
+            fprintf(output, "    mov rsi, [stack_top]\n");
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rcx, [rsi]\n");
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rax, [rsi]\n");
+            fprintf(output, "    shr rax, cl\n");
+            fprintf(output, "    mov [rsi], rax\n");
+            fprintf(output, "    add rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov [stack_top], rsi\n");
+        }
+        break;
+        case INST_SHL: {
+            fprintf(output, "    ;; shl\n");
+            fprintf(output, "    mov rsi, [stack_top]\n");
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rcx, [rsi]\n");
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rax, [rsi]\n");
+            fprintf(output, "    shl rax, cl\n");
+            fprintf(output, "    mov [rsi], rax\n");
+            fprintf(output, "    add rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov [stack_top], rsi\n");
+        }
+        break;
+        case INST_NOTB: {
+            fprintf(output, "    ;; notb\n");
+            fprintf(output, "    mov rsi, [stack_top]\n");
+            fprintf(output, "    sub rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rax, [rsi]\n");
+            fprintf(output, "    not rax\n");
+            fprintf(output, "    mov [rsi], rax\n");
+            fprintf(output, "    add rsi, BM_WORD_SIZE\n");
+            fprintf(output, "    mov [stack_top], rsi\n");
+        }
+        break;
+        case INST_READ8I: {
+            fprintf(output, "    ;; read8i\n");
+            fprintf(output, "    mov r11, [stack_top]\n");
+            fprintf(output, "    sub r11, BM_WORD_SIZE\n");
+            fprintf(output, "    mov rsi, [r11]\n");
+            fprintf(output, "    add rsi, memory\n");
+            fprintf(output, "    xor rax, rax\n");
+            fprintf(output, "    mov al, BYTE [rsi]\n");
+            fprintf(output, "    test al, al\n");
+            fprintf(output, "    jnl .read8i_%zu\n", jmp_count);
+            fprintf(output, "    neg al\n");
+            fprintf(output, "    neg rax\n");
+            fprintf(output, "    .read8i_%zu:\n", jmp_count);
+            fprintf(output, "    mov [r11], rax\n");
+            jmp_count += 1;
```

Comment:
Why read instruction would even need any branching?

Reply:
Because if value that was read is negative we need to move sing-bit to the beginning of `rax`
Without it situations [like that](https://cdn.discordapp.com/attachments/778284187849195550/837030233307742368/unknown.png) appear.

Reply:
Can you do `movsx %ax, %rax` ?

Reply:
Ok, so I removed branching in read instructions, but It's still used in `f2u` and `u2f` instructions